torch/csrc/autograd/python_saved_variable_hooks.h

```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+#include <pybind11/pybind11.h>
+#include <torch/csrc/autograd/python_variable.h>
+#include <torch/csrc/autograd/saved_variable_hooks.h>
+#include <torch/csrc/python_headers.h>
+#include <torch/csrc/THP_export.h>
+#include <ATen/ATen.h>
+
+namespace py = pybind11;
+
+namespace torch { namespace autograd {
+
+struct PySavedVariableHooks : public SavedVariableHooks {
+  PySavedVariableHooks(py::function &pack_hook, py::function &unpack_hook);
+  void call_pack_hook(at::Tensor &tensor) override;
+  at::Tensor call_unpack_hook() override;
+  ~PySavedVariableHooks() override;
+
+private:
+  PyObject* pack_hook_;
```

Comment:
Is the reason we can't store `py::function` directly that we need to leak it on destruction sometimes? It's interesting that pybind doesn't handle this case.

Reply:
No, we could actually properly release the py::functions in the destructor the same way.
The problem is that pybind is imported with visibility hidden (from the c++ definition) and so that means that we cannot use them as attributes on a class that is visible (due to the TORCH_API). You basically say that the class should be exposed but the attribute's type hidden which doesn't work.
Moving to PyObject (that is always exposed) is a simple fix for that. We might be able to make py::function exposed as well, but I don't know how from the top of my head.